duff.1
.\" Copyright (c) Camilla Berglund <elmindreda@users.sourceforge.net>
.\" This document is licensed under the zlib/libpng license.
.\" Please see the file `COPYING' for license details.
.\"
.Dd January 6, 2009
.Os
.Dt DUFF 1
.Sh NAME
.Nm duff
.Nd duplicate file finder
.Sh SYNOPSIS
.Nm
.Op Fl 0HLPaeqprtz
.Op Fl d Ar function
.Op Fl f Ar format
.Op Fl l Ar limit
.Op Ar
.Nm
.Op Fl h
.Nm
.Op Fl v
.Sh DESCRIPTION
The
.Nm
utility reports clusters of duplicates in the specified files and/or directories.
In the default mode,
.Nm
prints a customizable header, followed by the names of all the files in the cluster.
In excess mode,
.Nm
does not print a header, but instead for each cluster prints the names of all but the first of the files it includes.
.Pp
If no files are specified as arguments,
.Nm
reads file names from stdin.
.Pp
Note that as of version 0.4,
.Nm
ignores symbolic links to files, as that behavior was conceptually broken.
Therefore, the
.Fl H , L
and
.Fl P
options now apply only to directories.
.Pp
Also note that as of version 0.5,
.Nm
works on both regular files and block device nodes.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl 0
If reading file names from stdin, assume they are null-terminated, instead of separated by newlines.
Also, when printing file names and cluster headers, terminate them with null characters instead of newlines.
.Pp
This is useful for file names containing whitespace or other non-standard characters.
.It Fl H
Follow symbolic links listed on the command line.
This overrides any previous
.Fl L
or
.Fl P
option.
Note that this only applies to directories, as symbolic links to files are never followed.
.It Fl L
Follow all symbolic links.
This overrides any previous
.Fl H
or
.Fl P
option.
Note that this only applies to directories, as symbolic links to files are never followed.
.It Fl P
Don't follow any symbolic links.
This overrides any previous
.Fl H
or
.Fl L
option.
This is the default.
Note that this only applies to directories, as symbolic links to files are never followed.
.It Fl a
Include hidden files and directories when searching recursively.
.It Fl d Ar function
The message digest function to use.
The supported functions are 
.Ar sha1 , sha256 , sha384
and
.Ar sha512 .
The default is
.Ar sha1 .
.It Fl e
Excess mode. List all excess files; i.e. all but the first one encountered of each cluster of duplicates.
This is useful when you want to automate removal of duplicate files.
.It Fl f Ar format
Set the format of the cluster header.
If the header is set to the empty string, no header line is printed.
.Pp
The following escape sequences are available:
.Bl -hang -offset left
.It Cm %n
The number of files in the cluster.
.It Cm %c
A legacy synonym for
.Cm %d ,
for compatibility reasons.
.It Cm %d
The message digest of files in the cluster.
.It Cm %i
The one-based index of the file cluster.
.It Cm %s
The size, in bytes, of a file in the cluster.
.It Cm %%
A
.Sq %
character.
.El
.Pp
The default format string is:
.Pp
.Dl %n files in cluster %i (%s bytes, digest %d)
.It Fl h
Display help information and exit.
.It Fl l Ar limit
The minimum size of files to be sampled.
If the size of files in a cluster is equal or greater than the specified limit,
.Nm
will sample and compare a few bytes from each file before calculating a full digest.
This is merely an optimization for larger files and does not affect which files are considered by duff.
The default limit is 1048576 bytes.
.It Fl q
Quiet mode.
Suppress warnings and error messages.
.It Fl p
Physical mode.
Make
.Nm
consider physical files instead of hard links.
If specified, multiple hard links to the same physical file will not be reported as duplicates.
.It Fl r
Recursively search into all specified directories.
.It Fl t
Thorough mode.
Distrust digests as a guarantee for equality.
In thorough mode,
.Nm
compares files byte by byte when their digests match.
.It Fl v
Display version information and exit.
.It Fl z
Do not consider files with zero size to be equal.
This option suppresses reporting a cluster of empty files.
.El
.Sh EXAMPLES
.\" TODO: Fix the formatting of the example commands.
The command:
.Dl duff -e * \&| xargs rm
.Pp
removes all duplicate files in the current directory.
Note that it's usually a good idea to check what will be removed before running a command like the one above.
.Pp
The command:
.Dl duff -r foo/
.Pp
lists all duplicate files in the directory foo and its subdirectories.
.Pp
The command:
.Dl find \&. -name '*.h' -print0 \&| duff -0
.Pp
lists all duplicate header files in the current directory and its subdirectories, correctly handling file names containing whitespace.
.Sh DIAGNOSTICS
.Ex -std
.Sh "SEE ALSO"
.Xr find 1 ,
.Xr xargs 1
.Sh AUTHORS
.An "Camilla Berglund" Aq elmindreda@users.sourceforge.net
.Sh BUGS
.Nm
doesn't check whether the same file or block device node has been specified twice on the command line.
This may lead it to report single files or block device nodes as duplicates.
Note that this problem does not affect directories.
.Pp
.Nm
no longer (as of version 0.4) considers symbolic links to files potential duplicates, and will not report them.
This may break scripts relying on the previous behavior.
.Pp
If the underlying files are modified while duff is running, all bets are off.
This is not really a bug, but it can still bite you.


NEWS
User-visible changes in duff
============================

Version 0.5
-----------

* Added flag to handle null-terminated file names (-0).
* Added flag to select message digest function (-d).
* Improved handling of symbolic links to directories.
* Added internationalization via GNU gettext.
* Fixed a number of bugs

Version 0.4
-----------

* Added file name reading from stdin.
* Added primitive directory loop detection.
* Compatibility with more systems.
* Added physical mode flag (-p).
* Added argument symbolic link flag (-H).
* Improved warnings for sample limit flag (-l).
* Removed collection of symbolic links to files.
* Added reporting of access-protected empty files.
* Added example script for hard-linking duplicate files.
* Updated and improved manpage.
* Fixed a number of bugs.

Version 0.3.1
-------------

* Added flag to ignore empty files (-z).
* Fixed a number of bugs.

Version 0.3
-----------

* Added explicit link handling flags (-L,-P).
* Added sha1-asaddi for SHA1 checksum calculation.
* Completed basic manpage.
* Added sample size limit flag (-l).
* Added very brief hacking guide.
* Compatibility, portability, refactoring, cleanup.
* Added basic (non symlink-aware) excess mode (-e).

Version 0.2
-----------

* Added customisable cluster headers (-f).
* Split code into multiple files.
* Got SourceForge.net account (duff.sf.net).
* Added recursive flag (-r).
* Added quiet flag (-q).
* Started writing manpage.

Version 0.1
-----------

* Basic functionality.



duff.c
/*
 * duff - Duplicate file finder
 * Copyright (c) 2005 Camilla Berglund <elmindreda@users.sourceforge.net>
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any
 * damages arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any
 * purpose, including commercial applications, and to alter it and
 * redistribute it freely, subject to the following restrictions:
 *
 *  1. The origin of this software must not be misrepresented; you
 *     must not claim that you wrote the original software. If you use
 *     this software in a product, an acknowledgment in the product
 *     documentation would be appreciated but is not required.
 *
 *  2. Altered source versions must be plainly marked as such, and
 *     must not be misrepresented as being the original software.
 *
 *  3. This notice may not be removed or altered from any source
 *     distribution.
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif

/* Macros to get 64-bit off_t
 */
#define _GNU_SOURCE 1
#define _FILE_OFFSET_BITS 64

#if HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif

#if HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif

#if HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif

#if HAVE_INTTYPES_H
#include <inttypes.h>
#elif HAVE_STDINT_H
#include <stdint.h>
#endif

#if HAVE_ERRNO_H
#include <errno.h>
#endif

#if HAVE_UNISTD_H
#include <unistd.h>
#endif

#if HAVE_STDIO_H
#include <stdio.h>
#endif

#if HAVE_STRING_H
#include <string.h>
#endif

#if HAVE_STDARG_H
#include <stdarg.h>
#endif

#if HAVE_STDLIB_H
#include <stdlib.h>
#endif

#if HAVE_LIMITS_H
#include <limits.h>
#endif

#if HAVE_LOCALE_H
#include <locale.h>
#endif

#include "duffstring.h"
#include "duff.h"

/* The 'symlink dereference' mode.  Controls the handling of symlinks to
 * directories.  The different modes are defined in duff.h.
 */
SymlinkMode follow_links_mode = NO_SYMLINKS;
/* The 'all files' flag.  Includes dotfiles when searching recursively.
 */
int all_files_flag = 0;
/* The 'verbose' flag.  Makes the program verbose.
 */
int verbose_flag = 0;
/* The 'recursive' flag.  Recurses into all specified directories.
 */
int recursive_flag = 0;
/*! The 'null termination' flag.  Reads and writes null-terminated file
 *  names.
 */
int null_terminate_flag = 0;
/* The 'shut up' flag.  Makes the program not complain about skipped
 * non-files.
 */
int quiet_flag = 0;
/* The 'physical mode' flag.  Makes the program consider entries being
 * physical files instead of hard links.
 */
int physical_flag = 0;
/* The 'excess mode' flag.  For each duplicate cluster, reports all but one.
 * Useful for `xargs rm'.
 */
int excess_flag = 0;
/* The 'paranoid' flag.  Makes the program distrust message digests, forcing
 * byte-by-byte comparisons.
 */
int thorough_flag = 0;
/* The ignore empty files' flag.  Makes the program not report empty
 * files as duplicates.
 */
int ignore_empty_flag = 0;
/* The 'header format' value.  Specifies the look of the cluster header.
 * If set to the empty string, no headers are printed.
 */
const char* header_format = "%n files in cluster %i (%s bytes, digest %d)";
/* The 'sample limit' value.  Specifies the minimal size of files to be
 * compared with the sampling method.
 */
off_t sample_limit = 1048576;
/* The message digest function to use.
 */
enum Function digest_function = SHA_1;

/* These functions are documented below, where they are defined.
 */
static void version(void);
static void usage(void);
static void bugs(void);

/* Prints version information to stdout.
 */
static void version(void)
{
  printf("%s\n", PACKAGE_STRING);
  printf(gettext("Copyright (c) 2005 Camilla Berglund <elmindreda@users.sourceforge.net>\n"));
  printf(gettext("%s contains shaX-asaddi\n"), PACKAGE_NAME);
  printf(gettext("Copyright (c) 2001-2003 Allan Saddi <allan@saddi.com>\n"));
}

/* Prints brief help information to stdout.
 * Note that it is a good idea to keep this synchronised with the actual code.
 * Note that it is also a good idea it keep it synchronised with the manpage.
 */
static void usage(void)
{
  /* TODO: Internationalize this */

  printf("usage: %s [-0HLPaepqrtz] [-d function] [-f format] [-l size] [file ...]\n", PACKAGE_NAME);
  printf("       %s -h\n", PACKAGE_NAME);
  printf("       %s -v\n", PACKAGE_NAME);
  printf("options:\n"
         "  -0  read and write file names terminated by a null character\n"
         "  -H  follow symbolic links on the command line\n"
         "  -L  follow all symbolic links\n"
         "  -P  do not follow any symbolic links (default)\n"
         "  -a  all files; include hidden files when searching recursively\n"
	 "  -d  the message digest function to use\n"
         "  -e  excess files mode; list only excess files (no headers)\n"
         "  -f  header format; set format for cluster headers\n"
         "  -h  show this help\n"
         "  -l  size limit; the minimal size that activates sampling\n"
         "  -q  quiet; suppress warnings and error messages\n"
         "  -p  physical mode; do not report multiple links\n"
         "  -r  recursive; search in specified directories\n"
         "  -t  thorough; force byte-by-byte comparison of files\n"
         "  -v  show version information\n"
         "  -z  do not report empty files\n");
}

/* Prints bug report address to stdout.
 */
static void bugs(void)
{
  printf(gettext("Report bugs to <%s>\n"), PACKAGE_BUGREPORT);
}

/* I don't know what this function does.
 * I just put it in because it looks cool.
 */
int main(int argc, char** argv)
{
  int i, ch;
  char* temp;
  off_t limit;
  char path[PATH_MAX];

  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);

  while ((ch = getopt(argc, argv, "0HLPad:ef:hl:pqrtvz")) != -1)
  {
    switch (ch)
    {
      case '0':
	null_terminate_flag = 1;
	break;
      case 'H':
	follow_links_mode = ARG_SYMLINKS;
	break;
      case 'L':
	follow_links_mode = ALL_SYMLINKS;
	break;
      case 'P':
	follow_links_mode = NO_SYMLINKS;
	break;
      case 'a':
        all_files_flag = 1;
        break;
      case 'd':
	if (strcmp(optarg, "sha1") == 0)
	  digest_function = SHA_1;
	else if (strcmp(optarg, "sha256") == 0)
	  digest_function = SHA_256;
	else if (strcmp(optarg, "sha384") == 0)
	  digest_function = SHA_384;
	else if (strcmp(optarg, "sha512") == 0)
	  digest_function = SHA_512;
	else
	  error(gettext("%s is not a supported digest function"), optarg);
	break;
      case 'e':
        excess_flag = 1;
	break;
      case 'f':
        header_format = optarg;
        break;
      case 'h':
        usage();
        bugs();
        exit(0);
      case 'l':
        limit = (off_t) strtoull(optarg, &temp, 10);
	if (temp == optarg || errno == ERANGE || errno == EINVAL)
	  warning(gettext("Ignoring invalid sample limit %s"), optarg);
	else
	{
	  if (limit < SAMPLE_COUNT)
	    warning(gettext("Sample limit must be at least %u bytes"), SAMPLE_COUNT);
	  else
	    sample_limit = limit;
	}
	break;
      case 'p':
	physical_flag = 1;
	break;
      case 'q':
        quiet_flag = 1;
        break;
      case 'r':
        recursive_flag = 1;
        break;
      case 't':
        thorough_flag = 1;
        break;
      case 'v':
        version();
        exit(0);
      case 'z':
	ignore_empty_flag = 1;
	break;
      default:
        usage();
        bugs();
        exit(1);
    }
  }
  
  argc -= optind;
  argv += optind;

  if (argc)
  {
    /* Read file names from command line */
    for (i = 0;  i < argc;  i++)
    {
      kill_trailing_slashes(argv[i]);
      process_path(argv[i], 0);
    }
  }
  else
  {
    /* Read file names from stdin */
    while (read_path(stdin, path, sizeof(path)) == 0)
    {
      kill_trailing_slashes(path);
      process_path(path, 0);
    }
  }
  
  report_clusters();
  
  exit(0);
}

